### ISSUE TITLE
Query builder 'where' does not work with russian words

### ISSUE THREAD

**Issue by siarheipashkevich** [laravel/framework#42816]
- Laravel Version: 8.83.16
- PHP Version: 8.1.7
- Database Driver & Version: MySQL 8.0.29

### Description:

Hi, all.

I'm working with laravel a few years but today I'm facing with strange situation with query builder which does not work correctly.

### Steps To Reproduce:

1. Create combined query:
https://github.com/siarheipashkevich/query-builder/blob/54f58dc93bd4474293c6fa5823f7d09f891348f5/tests/Feature/ExampleTest.php#L45
2. Use `like` operator in `where` method:
https://github.com/siarheipashkevich/query-builder/blob/master/tests/Feature/ExampleTest.php#L51

I've created repository with test for this situation and expected result:
https://github.com/siarheipashkevich/query-builder

Please clone and run tests.

**Comment by ARehmanMahi** [comment]
What do you mean doesn't not work. Does it error out and gives error, unexpected results?
Can you show what you expect and what you get?

**Comment by siarheipashkevich** [comment]
@ARehmanMahi I created repository with test which has expected result.

**Comment by ARehmanMahi** [comment]
psr/log dependency requires PHP 8.0, But my system isn't configured with that. Hope someone else chimes in. Cheers

**Comment by rodrigopedra** [comment]
You tests are configured to run with sqlite.

And with sqlite console, the query returns no records too:

~~~
$ sqlite3 database/database.sqlite 
SQLite version 3.38.5 2022-05-06 15:25:27
Enter ".help" for usage hints.
sqlite> select name from dishes;
Kari Schmeler
Ms. Daphnee Trantow DDS
Josefina Lehner
Devyn Vandervort
Gilda Bednar
Mr. Mitchell Schaden IV
Marion Hickle
Kaia Connelly I
Elinor Bode
Queenie Bradtke IV
Тест 1
Test 1
sqlite> select name from dishes where '%тест%';
sqlite> .quit
~~~

Can you try setting your tests to run with another database?

Otherwise, it is either a SQLite limitation, configuration, or maybe it requires a collation to be specified when querying with no latin characters.

If you believe this is a bug, it should be reported to SQLite, as there is not much Laravel can do if the underlying database engine does not work with this.

**Comment by siarheipashkevich** [comment]
It's reproduced for MySQL 8.0.29 also, in tests I'm using sqlite and also reproduced..

**Comment by siarheipashkevich** [comment]
@ARehmanMahi can you try `select name from dishes where name like '%Тест%';`

**Comment by rodrigopedra** [comment]
Sure:

~~~
$ sqlite3 database/database.sqlite 
SQLite version 3.38.5 2022-05-06 15:25:27
Enter ".help" for usage hints.
sqlite> select name from dishes;
Richard Nikolaus I
Dimitri Raynor III
Patience Hilpert
Andreanne Prosacco
Camron Stroman MD
Miss Ivy Monahan IV
Prof. Reina Osinski
Ms. Alexanne Kerluke
Mrs. Hellen Rosenbaum PhD
Kelvin Ortiz
Тест 1
Test 1
sqlite> select name from dishes where name like '%Тест%';
Тест 1
sqlite> select name from dishes where name like '%test%';
Test 1
sqlite> .quit
~~~

I changed your test commenting the Russian where clause and uncommenting the English one, and the assert also fails when using MySQL.

Can you try this test case instead, with MySQL:

~~~php
<?php

namespace Tests\Feature;

use App\Models\{Dish, ExcludedItem, Preparation, User};
use Illuminate\Database\Query\JoinClause;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Illuminate\Support\Facades\DB;
use Tests\TestCase;

class ExampleTest extends TestCase
{
    use RefreshDatabase;

    /**
     * A basic test example.
     *
     * @return void
     */
    public function test_example()
    {
        $user = User::factory()->create();

        Dish::factory()->count(10)->create();
        Preparation::factory()->count(10)->create();

        $excludedDishRus = Dish::factory()->create(['name' => 'Тест 1']);
        $excludedPreparationRus = Preparation::factory()->create(['name' => 'Тест 2']);
        $excludedDishEn = Dish::factory()->create(['name' => 'Test 1']);
        $excludedPreparationEn = Preparation::factory()->create(['name' => 'Test 2']);

        $user->excludedItems()->create(['item_id' => $excludedDishRus->id, 'item_type' => 'dish']);
        $user->excludedItems()->create(['item_id' => $excludedPreparationRus->id, 'item_type' => 'preparation']);
        $user->excludedItems()->create(['item_id' => $excludedDishEn->id, 'item_type' => 'dish']);
        $user->excludedItems()->create(['item_id' => $excludedPreparationEn->id, 'item_type' => 'preparation']);

        $dishesQuery = Dish::query()
            ->select(['dishes.id', 'dishes.name', DB::raw("'dish' as item_type")]);

        $preparationsQuery = Preparation::query()
            ->select(['preparations.id', 'preparations.name', DB::raw("'preparation' as item_type")]);

        $itemsQuery = $dishesQuery->union($preparationsQuery);

        $countWithoutFilter = ExcludedItem::query()
            ->joinSub($itemsQuery, 'items', function (JoinClause $join) {
                $join->on('items.id', 'excluded_items.item_id')
                    ->on('items.item_type', 'excluded_items.item_type');
            })
            ->where('items.name', 'like', '%тест%')
            ->select(['excluded_items.*', 'items.name'])
            ->count();

        $this->assertEquals(2, $countWithoutFilter);
    }
}
~~~

Note I am not querying from the user relation.

It seems there is a problem with the query when querying from the user model.

**Comment by rodrigopedra** [comment]
This is working, with MySQL:

~~~php
<?php

namespace Tests\Feature;

use App\Models\{Dish, Preparation, User};
use Illuminate\Database\Query\JoinClause;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Illuminate\Support\Facades\DB;
use Tests\TestCase;

class ExampleTest extends TestCase
{
    use RefreshDatabase;

    /**
     * A basic test example.
     *
     * @return void
     */
    public function test_example()
    {
        $user = User::factory()->create();

        Dish::factory()->count(10)->create();
        Preparation::factory()->count(10)->create();

        $excludedDishRus = Dish::factory()->create(['name' => 'Тест 1']);
        $excludedPreparationRus = Preparation::factory()->create(['name' => 'Тест 2']);
        $excludedDishEn = Dish::factory()->create(['name' => 'Test 1']);
        $excludedPreparationEn = Preparation::factory()->create(['name' => 'Test 2']);

        $user->excludedItems()->create(['item_id' => $excludedDishRus->id, 'item_type' => 'dish']);
        $user->excludedItems()->create(['item_id' => $excludedPreparationRus->id, 'item_type' => 'preparation']);
        $user->excludedItems()->create(['item_id' => $excludedDishEn->id, 'item_type' => 'dish']);
        $user->excludedItems()->create(['item_id' => $excludedPreparationEn->id, 'item_type' => 'preparation']);

        $dishesQuery = Dish::query()
            ->select(['dishes.id', 'dishes.name', DB::raw("'dish' as item_type")])
            ->where('name', 'like', '%тест%');

        $preparationsQuery = Preparation::query()
            ->select(['preparations.id', 'preparations.name', DB::raw("'preparation' as item_type")])
            ->where('name', 'like', '%тест%');

        $itemsQuery = $dishesQuery->union($preparationsQuery);

        $countWithoutFilter = $user->excludedItems()
            ->joinSub($itemsQuery, 'items', function (JoinClause $join) {
                $join->on('items.id', 'excluded_items.item_id')
                    ->on('items.item_type', 'excluded_items.item_type');
            })
            ->select(['excluded_items.*', 'items.name'])
            ->count();

        $this->assertEquals(2, $countWithoutFilter);
    }
}
~~~

Note I moved the where clause to each "unioned" query

**Comment by rodrigopedra** [comment]
@siarheipashkevich it seems SQLite `LIKE` operator is case-insensitve only for ASCII characters, and in SQLite `UPPER(...)` also only seems to work for ASCII characters.

For SQLite you can register custom functions, so this should work with **SQLite**:

~~~php
<?php

namespace Tests\Feature;

use App\Models\{Dish, Preparation, User};
use Illuminate\Database\Query\JoinClause;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Str;
use Tests\TestCase;

class ExampleTest extends TestCase
{
    use RefreshDatabase;

    /**
     * A basic test example.
     *
     * @return void
     */
    public function test_example()
    {
        $user = User::factory()->create();

        Dish::factory()->count(10)->create();
        Preparation::factory()->count(10)->create();

        $excludedDishRus = Dish::factory()->create(['name' => 'Тест 1']);
        $excludedPreparationRus = Preparation::factory()->create(['name' => 'Тест 2']);
        $excludedDishEn = Dish::factory()->create(['name' => 'Test 1']);
        $excludedPreparationEn = Preparation::factory()->create(['name' => 'Test 2']);

        $user->excludedItems()->create(['item_id' => $excludedDishRus->id, 'item_type' => 'dish']);
        $user->excludedItems()->create(['item_id' => $excludedPreparationRus->id, 'item_type' => 'preparation']);
        $user->excludedItems()->create(['item_id' => $excludedDishEn->id, 'item_type' => 'dish']);
        $user->excludedItems()->create(['item_id' => $excludedPreparationEn->id, 'item_type' => 'preparation']);

        $dishesQuery = Dish::query()
            ->select(['dishes.id', 'dishes.name', DB::raw("'dish' as item_type")]);

        $preparationsQuery = Preparation::query()
            ->select(['preparations.id', 'preparations.name', DB::raw("'preparation' as item_type")]);

        $itemsQuery = $dishesQuery->union($preparationsQuery);

        DB::connection()->getPdo()->sqliteCreateFunction('UTF8_UPPER', function (string $value) {
            return Str::upper($value);
        }, 1);

        $countWithoutFilter = $user->excludedItems()
            ->joinSub($itemsQuery, 'items', function (JoinClause $join) {
                $join->on('items.id', 'excluded_items.item_id')
                    ->on('items.item_type', 'excluded_items.item_type');
            })
            ->whereRaw('UTF8_UPPER(name) like ?', Str::upper('%тест%'))
            ->select(['excluded_items.*', 'items.name'])
            ->count();

        $this->assertEquals(2, $countWithoutFilter);
    }
}
~~~

Note the custom registered function.

This is a good example why we should run our tests in the same database we use for our project.

**Comment by siarheipashkevich** [comment]
I've dumped query in test:
![Screenshot from 2022-06-15 18-35-22](https://user-images.githubusercontent.com/7753600/173867805-3d170c73-341c-4a2f-89f5-787af3e129a5.png)

```
select `excluded_items`.*, `items`.`name` from `excluded_items` inner join ((select `dishes`.`id`, `dishes`.`name`, 'dish' as item_type from `dishes`) union (select `preparations`.`id`, `preparations`.`name`, 'preparation' as item_type from `preparations`)) as `items` on `items`.`id` = `excluded_items`.`item_id` and `items`.`item_type` = `excluded_items`.`item_type` where `excluded_items`.`user_id` = 1 and `excluded_items`.`user_id` is not null and `items`.`name` like '%Тест%'
```

And if run this query in mysql I'm getting expected result:
![Screenshot from 2022-06-15 18-37-54](https://user-images.githubusercontent.com/7753600/173868355-fb79e769-e1f5-4840-a4d5-73fe3a698126.png)


But in laravel application it is empty. Why?

**Comment by rodrigopedra** [comment]
If you use MySQL for tests, and if you use the string `'%Тест%'` as you did in your screenshot above, tests do pass. In other words, records are returned.

And if you run the query in MySQL if the string `'%тест%'`, as it is in your test repository, the query returns empty, both in MySQL and in the test case, which ends up failing.

Problem seems to be with case sentitivity, which is not a Laravel issue, but rather a database issue.

Summary:

- Using `'%Тест%'` : If run in MySQL it returns 2 records, as expected. Laravel test passes, if using this string

https://github.com/rodrigopedra/query-builder/blob/5f09ba7380fa8598050fb4c09ac97f6e1cefa0d8/tests/Feature/ExampleTest.php#L46
 
- Using `'%тест%'` : If run in MySQL it returns no records, probably by case change. Laravel test fails, if using this string

https://github.com/rodrigopedra/query-builder/blob/5f09ba7380fa8598050fb4c09ac97f6e1cefa0d8/tests/Feature/ExampleTest.php#L86

I uploaded a commit to a fork of your test repo. Here is the link:

https://github.com/rodrigopedra/query-builder

After cloning run `composer install`, configure `.env.testing` with your MySQL database credentials, and run `php artisan test`.

~~~
$ php artisan test

   PASS  Tests\Unit\ExampleTest
  ✓ example

   FAIL  Tests\Feature\ExampleTest
  ✓ this will pass due to matching case
  ⨯ this will fail due to mismatch case

  ---
~~~

Note first test case passes, as casing are matching in both database and query

If you are using SQLite, both in tests and project, I already provided a solution to provide SQLite with a custom function. SQLite limitation is also has a link to their official site documentation.

If you are using MySQL, and want the query to be case insensitive, please read MySQL's documentation. I guess you can mangle with collation arguments to  get it working.

**Comment by siarheipashkevich** [comment]
About your notes for mysql `%Тест%` and `%тест%` returns results in both cases - it is valid.

**Comment by siarheipashkevich** [comment]
@ARehmanMahi I've tried your repository.

sqlite
![Screenshot from 2022-06-15 20-51-27](https://user-images.githubusercontent.com/7753600/173892819-9d8cc73c-9872-4d53-8d30-a922b247b726.png)

mysql
![Screenshot from 2022-06-15 20-51-16](https://user-images.githubusercontent.com/7753600/173892839-d707c5f4-9e2f-401b-af97-b3d0c2b843bd.png)

**Comment by rodrigopedra** [comment]
Yes, you are right. I had added the factory seeding to the `DatabaseSeeder` locally, so I could play with the data, and when restoring the changes for commit I might have made a confusion.

I also assumed `.env.testing` would override phpunit.xml automatically, but it doesn't work like that.

For some reason, when not querying from the user, e.g. without the user constraint, the LIKE operator works in Laravel.

With the user, if we move the WHEREs to the queries used in union, it also works with MySQL

~~~php
$ php artisan test --env=testing

   PASS  Tests\Unit\ExampleTest
  ✓ example

   PASS  Tests\Feature\ExampleTest
  ✓ this passes with matching case and changed wheres
  ✓ this passes with mismatching case and changed wheres
  ✓ this passes with mismatching case and no user

  Tests:  4 passed
  Time:   0.22s
~~~

I sent a new commit to my fork with the changes.

It surely is not consistent behavior, as running the original queries from MySQL yields different results. But it is beyond my knowledge.

I apologize for the confusion.

**Comment by siarheipashkevich** [comment]
Yes, it will work.
But why is not working in my original case using laravel way? Because I have many unions and don't want to add for each subquery condition.

**Comment by rodrigopedra** [comment]
I am sorry, I don’t know.

It seems to be either a bug, either in Laravel or PDO, or maybe some PDO configuratin can solve it.

But that is beyond my knowledge.

if you can test with using PDO directly, using the same configuration as Laravel uses, to see if it works as expected, we will at least know if the issue is with how PDO handle this query, or if it is with Laravel

**Comment by siarheipashkevich** [comment]
If I execute this queries in laravel using `DB::select('...')` then I get results also.
```
select `excluded_items`.*, `items`.`name` from `excluded_items` inner join ((select `dishes`.`id`, `dishes`.`name`, 'dish' as item_type from `dishes`) union (select `preparations`.`id`, `preparations`.`name`, 'preparation' as item_type from `preparations`)) as `items` on `items`.`id` = `excluded_items`.`item_id` and `items`.`item_type` = `excluded_items`.`item_type` where `excluded_items`.`user_id` = 1 and `excluded_items`.`user_id` is not null and `items`.`name` like '%Тест%'
```
```
select `excluded_items`.*, `items`.`name` from `excluded_items` inner join ((select `dishes`.`id`, `dishes`.`name`, 'dish' as item_type from `dishes`) union (select `preparations`.`id`, `preparations`.`name`, 'preparation' as item_type from `preparations`)) as `items` on `items`.`id` = `excluded_items`.`item_id` and `items`.`item_type` = `excluded_items`.`item_type` where `excluded_items`.`user_id` = 1 and `excluded_items`.`user_id` is not null and `items`.`name` like '%тест%'
```

**Comment by rodrigopedra** [comment]
It looks like a bug indeed. I created this test command:

~~~php
<?php // ./route/console.php

use App\Models\Dish;
use App\Models\Preparation;
use App\Models\User;
use Illuminate\Database\Query\JoinClause;
use Illuminate\Support\Facades\Artisan;
use Illuminate\Support\Facades\DB;

Artisan::command('eloquent', function () {
    $searchString = '%тест%';

    // baseline: proof MySQL works with search string
    $dishes = Dish::query()->where('name', 'like', $searchString)->get();
    \dump($dishes->toArray());

    // test
    $dishesQuery = Dish::query()
        ->select(['dishes.id', 'dishes.name', DB::raw("'dish' as item_type")]);

    $preparationsQuery = Preparation::query()
        ->select(['preparations.id', 'preparations.name', DB::raw("'preparation' as item_type")]);

    $itemsQuery = $dishesQuery->union($preparationsQuery);

    $user = User::query()->find(1);

    $builder = $user->excludedItems()
        ->joinSub($itemsQuery, 'items', function (JoinClause $join) {
            $join->on('items.id', 'excluded_items.item_id')
                ->on('items.item_type', 'excluded_items.item_type');
        })
        ->select(['excluded_items.*', 'items.name'])
        ->where('items.name', 'like', $searchString);

    $statement = $builder->getConnection()->getPdo()->prepare($builder->toSql());
    $statement->execute([1, $searchString]);

    \dd(
        $statement->fetchAll(\PDO::FETCH_ASSOC), // has results
        $builder->get()->toArray(), // empty results
    );
});
~~~

And even when using Eloquent's PDO instance, results are not the same.

I pushed this command to my fork, as well an updated `DatabaseSeeder` to make test data easier to generate.

I am looking into if it could be how `HasMany` relation object handles the query,  but I will have little to no time today to investigate this.

**Comment by siarheipashkevich** [comment]
@ARehmanMahi thank you so much for your help.

Potentially a problem in laravel builder?

**Comment by rodrigopedra** [comment]
@siarheipashkevich I found out if you add `PDO::ATTR_EMULATE_PREPARES => true` to your MySQL database options, it will work.

https://github.com/rodrigopedra/query-builder/blob/3b9a94adbd31d1607f43efbbc96e6e2492dc8f7a/config/database.php#L61-L64

It seems MySQL driver does not work well with some character sets and mixing `PDO::PARAM_INT`, `PDO::PARAM_STR`.

I still didn't find an official bug report on PHP, I found a hint on this on a forum (although there the user was having problem with Scandinavian characters).

If you comment out line 619 of `Illuminate\Database\Connection` class (as in 8.x branch you indicated you are using), it will work without that additional option:

https://github.com/laravel/framework/blob/246105598226ff590d51082f84e0ef04ab09d1fb/src/Illuminate/Database/Connection.php#L619

But I really doubt maintainers would accept a PR removing that line.

Although it seems to be a PHP bug, one could avoid it by not mixing parameter bindings when calling `$statement->bindValues()`, so I think it is worth investigating on Laravel side.

Maybe someone with more knowledge in PDO/MySQL native driver can jump in, and further investigate.

There is a somewhat related bug on PHP bug tracker, maybe it can help better find the root cause:

https://bugs.php.net/bug.php?id=79276
